### ISSUE TITLE
Deterministic cudnn algorithms

### ISSUE THREAD

**Issue by szagoruyko** [pytorch/pytorch#114]
It is totally possible to implement deterministic training with lua torch avoiding some cunn modules and cudnn algos. cutorch has atomicAdd in THIndexTensor which shouldn't pose a problem.
PyTorch should implement:
- ordered parallel data loading similar to torchnet https://github.com/torchnet/torchnet/blob/master/dataset/paralleldatasetiterator.lua#L53
- cudnn should allow user set deterministic mode somehow

**Comment by soumith** [comment]
More context by @daemon here:

https://github.com/pytorch/pytorch/issues/2831#issue-259826541

**Comment by soumith** [comment]
@prigoyal this one is for you.

Add a mode:

`torch.backends.cudnn.deterministic = True` which will enable CuDNN algorithms.

Here's some code pointers:
https://github.com/pytorch/pytorch/blob/master/torch/backends/cudnn/__init__.py
https://github.com/pytorch/pytorch/blob/master/torch/csrc/cudnn/Conv.cpp

**Comment by soumith** [comment]
CuDNN has deterministic flag now, and convolution algorithms are selected respecting this. See #2893

**Comment by B1azingB1ade** [comment]
Thanks for the fix. But my question is: Why not have `torch.backends.cudnn.deterministic = True` by default? I noticed that I got non-deterministic results and it took some effort to dig up this solution. Imagine someone who doesn't know about this and run some expensive simulations for research only to have results not reproducible. 
I feel that `torch.backends.cudnn.deterministic` should at least be set to `True` automatically whenever `torch.cuda.manual_seed()` is called, because determinism is the point of setting seed. From my testing, `torch.backends.cudnn.deterministic = True` doesn't noticeably slow down the performance, so are there reasons that we don't want the default to be `True`? Thanks!

**Comment by apaszke** [comment]
Deterministic algorithms can be dramatically slower than the fastest ones. It's just they for some reason your problem sizes didn't trigger those specialized paths, but others would suffer.

**Comment by B1azingB1ade** [comment]
I see. It's probably model-dependent. For reference I was running a model similar to the one from [Action-Conditional Video Prediction using Deep Networks in Atari Games](https://arxiv.org/abs/1507.08750). If I set `torch.backends.cudnn.enabled = False` the training time is about twice as long; if instead I set `torch.backends.cudnn.deterministic = True` the time difference is within 1%. I'll certainly keep the speed concern in mind when I run other models. Thanks.

**Comment by themightyoarfish** [comment]
Please consider adding documentation for this. I've been googling for some time and the only way of finding this seemed to be actually asking here on GitHub. Or maybe StackOverflow.

**Comment by ezyang** [comment]
It seems reasonable to add references to this flag from all functions which could dispatch to cuDNN. If you submit a PR this ping me and I'll merge it.

**Comment by themightyoarfish** [comment]
How do I figure out which functions this would apply to? I believe a good alternative would be to add a section on reproducibility to the Sphinx doc which then shows up when searching for "randomness", "deterministic" or something like that.

**Comment by ezyang** [comment]
AFAICT, it's only convolution and ctc_loss

```
[ezyang@devvm884.prn3 ~/local/eden-fbsource-tmp/fbcode/caffe2] ag deterministicCuDNN aten/
aten/src/ATen/Context.cpp
63:bool Context::deterministicCuDNN() const {

aten/src/ATen/Context.h
108:  bool deterministicCuDNN() const;

aten/src/ATen/native/Convolution.cpp
284:                          ctx.benchmarkCuDNN(), ctx.deterministicCuDNN(), ctx.userEnabledCuDNN());

aten/src/ATen/native/LossCTC.cpp
352:    res = std::get<0>(at::_cudnn_ctc_loss(log_probs, targets, input_lengths, target_lengths, BLANK, ctx.deterministicCuDNN()));
```

**Comment by themightyoarfish** [comment]
That entails a note should be added to all `torch.nn.Conv*` modules, yes?

**Comment by ezyang** [comment]
Yes, that's a reasonable start.

**Comment by kwea123** [comment]
Does CUDA version or CUDNN version affect the determinism?

**Comment by ezyang** [comment]
Across versions I don't think you can expect values to be equal, no (as they may have changed algorithms)
